docs/proposals/deployment.md

```diff
@@ -0,0 +1,172 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/deployment.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Deployment
+
+## Abstract
+
+A proposal for implementing a new resource - Deployment - which will enable
+declarative config updates for Pods and ReplicationControllers.
+
+Users will be able to create a Deployment, which will spin of
+a ReplicationController to bring up the desired pods.
+Users can also target the Deployment at existing ReplicationControllers, in
+which case the new RC will replace the existing ones. The exact mechanics of
+replacement depends on the DeploymentStrategy chosen by the user.
+DeploymentStrategies are explained in detail in a later section.
+
+## Implementation
+
+### API Object
+
+The `Deployment` API object will have the following structure:
+
+```go
+type Deployment struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior of the Deployment.
+  Spec DeploymentSpec
+
+  // Most recently observed status of the Deployment.
+  Status DeploymentStatus
+}
+
+type DeploymentSpec struct {
+  // Number of desired pods. This is a pointer to distinguish between explicit
+  // zero and not specified. Defaults to 1.
+  Replicas *int
+
+  // Label selector for pods. Existing ReplicationControllers which have all these
+  // label selectors will be scaled down.
+  Selector map[string][string]
+
+  // Describes the pods that will be created.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
+
+  // Rolling update config params. Present only if DeploymentType =
+  // RollingUpdate.
+  RollingUpdate *RollingUpdateDeploymentSpec
+}
+
+type DeploymentType string
+
+const (
+  // Kill all existing pods before creating new ones.
+  DeploymentRecreate DeploymentType = "Recreate"
+
+  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The percentage of pod replicas that should be updated in each interval.
+  // If nil, a fixed value of 1 pod is used.
+  UpdatePercent *int
```

Comment:
After implementing this and coming back today to review the docs here, I think I might be unclear on one nuance. For this case:

>maxUnavailabilityPercentage=10 and maxSurgePercentage = 20: The size of the new RC would be set to 30% of the original/desired immediately. The old RC would be downsized by 10% immediately. Progress thereafter would proceed as in the cases above.

Disregarding readiness (i.e., assume that after a scale up, all pods are immediately available):

Iteration 1: scale up 30%, scale down 10%
Iteration 2: scale up 30%, scale down 10%

or...

Iteration 2: scale up 10%, scale up 10%

In other words, does the surge maxium apply to every interval, or just the initial scale up (after which point only the unavailability max is considered)? My thought is that since maxSurgePercentage >0 means go up/down (rather than down/up), you have the extra capacity to do so and so want scale ups to occur faster than scale downs throughout the entire update rather than a single initial burst.